rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala

```diff
@@ -93,22 +117,27 @@ object ChargingRSpace {
       ): F[Unit] =
         result match {
 
-          case None => Sync[F].unit
+          case None => charge[F](eventStorageCost(triggeredBy.channelsCount))
 
           case Some((cont, dataList)) =>
-            val refundForConsume =
-              if (cont.persistent && triggeredBy != Consume) Cost(0)
-              else
-                storageCostConsume(cont.channels, cont.patterns, cont.value)
-
-            val refundForProduces = refundForRemovingProduces(
-              dataList,
-              cont.channels,
-              triggeredBy
-            )
+            for {
+              id <- consumeId(cont.value)
 
-            charge[F](Cost(-refundForConsume.value, "consume storage refund")) >>
-              charge[F](Cost(-refundForProduces.value, "produces storage refund"))
+              // We refund for non-persistent continuations, and for the persistent continuation triggering the comm.
+              // That persistent continuation is going to be charged for (without refund) once it has no matches in TS.
+              refundForConsume = if (!cont.persistent || id == triggeredBy.id) {
+                storageCostConsume(cont.channels, cont.patterns, cont.value)
+              } else {
+                Cost(0)
+              }
+              refundForProduces = refundForRemovingProduces(dataList, cont.channels, triggeredBy)
+
+              _ <- charge[F](Cost(-refundForConsume.value, "consume storage refund"))
+              _ <- charge[F](Cost(-refundForProduces.value, "produces storage refund"))
+              _ <- charge[F](eventStorageCost(triggeredBy.channelsCount))
```

Comment:
As I was saying in standup, the event log is emptied upon each call to `createCheckpoint`, which is done after each batch of deploys is processed. Why are you charging for this?

Reply:
> It doesn't make sense to charge a non-refundable fee for data that will certainly be removed

The event log is going to be stored in the BlockMessage, and will never be removed. At least not until we implement finalization and dag pruning based on it.

> difficult for users to see why their program costs what it does.

It's actually easier to estimate than with tuplespace storage: you pay 128 phlo per COMM, and 64 per produce/consume (assuming single-channel consumes).

Reply:
> The event log is going to be stored in the BlockMessage, and will never be removed.

If you're in favor of charging for the space occupied by deploy information, then you must also be in favor of charging for space occupied by the source code and all other information stored in `ProcessedDeploy`.

I see the problem you're trying to solve, but I think a better solution is to attach an execution cost to produces and consumes. Firstly, programs like `x!!(Q) | for(_ <= x){ P }` would actually terminate. Secondly, since there would always be a cost associated with produces and consumes (which are the only ways to add entries to the event logs), agents would not be able to create arbitrary event log entries without recourse.

Reply:
> then you must also be in favor of charging for space occupied by the source code and all other information stored in ProcessedDeploy.

At least the part that has size dependent on the deploy. I think we're doing so, but feel free to inspect that and create relevant JIRA tickets.

> a better solution is to attach an execution cost to produces and consumes

Any storage cost also doubles as execution cost. One has to process the data to store it.

> programs like x!!(Q) | for(_ <= x){ P } would actually terminate

They do under the current approach, so this is not a differentiator.

> agents would not be able to create arbitrary event log entries without recourse.

This is also true under the current approach, as adding arbitrary log entries will cause replay failures.

The current approach has a benefit of being more directly bound to a physical measure of cost (byte size) vs using an arbitrary iteration cost. The latter would miss the fact that iteration cost is different for a multi-channel receive than for a single-channel one. So: event-storage-size is more accurate.

Reply:
I'll reiterate:

Any storage cost also doubles as execution cost. One has to process the data to store it. 

The size-based (execution/storage)-cost is more accurate than adding a constant on each iteration, because it's less detached from actual work that needs to be done.

> not differentiate space complexity and time complexity, and that is absolutely wrong

It is, and that's what we have currently with TS storage costs. Changing that is out of this PR's scope. Any changes to that would need to acknowledge both 'storage cost as CPU cost' and 'storage cost per se', so it makes sense to use event storage cost for now.

> The time-complexity of produces and consumes is not equivalent to their space-complexity

:+1: At the same time, any space-cost not refunded for during the deploy also implies a time-cost (serialization).

> the time-cost of a persistent produce/consume will be the iterated time-cost of an ephemeral produce/consume

The time-cost of (whatever) produce/consume is lower-bounded by its event storage cost. I've not seen a more precise proposal from you, and anything correct you could propose would need to account for that. So feel free to file a JIRA describing the intended improvement.

> Regarding using cost accounting to compensate validators for system storage (...)

This whole section applies equally well to the TS storage costs we have now, and improving that is also out of scope of this PR.

> In general, your proposal to use cost accounting to compensate validators for resources consumed as a result of our protocol opens the door to a large investigation (...)

which is out of scope of this PR.

I consider this resolved. You seem to not acknowledge this, but the event storage costs are doing what you've asked for, and more.